### PR TITLE
[9.x] Improve redis test suite

### PR DESCRIPTION
**Pull request by TheLevti**
This is a follow up of https://github.com/laravel/framework/pull/40569, splitting the original pull request up. This pull request is about refining the existing redis test suite making it easy to test future changes related to redis.

### What changed

- Added data providers with named data sets to allow easy testing of different redis connection configurations.
- All tests in all components that use redis are now using the new data providers to test all supported redis connection settings separately making it easy to pinpoint which redis setting causes an issue with an existing or new component.

If any test related to redis fails, it will now output the exact data set that caused the failure:

image

### How to use

Now when you need to test something where redis is used, you can simply add the helper trait in the beginning like so:

```php
use InteractsWithRedis;
```

Then to let a test run through all possible redis variations, add one of the following doc blocks:

```php
/**
 * Runs this test with only two basic redis connections where one is using the predis driver and
 * the other the phpredis driver.
 *
 * @dataProvider redisConnectionDataProvider
 */
public function testSomethingRelatedToRedis($connection)
{
    $this->app['redis'] = $this->getRedisManager($connection);
    ...
}
```

```php
/**
 * Runs this test with the basic connection set and in addition advanced connection configurations like
 * persistent connections, prefix, scan options, serialization, compression etc.
 *
 * @dataProvider extendedRedisConnectionDataProvider
 */
public function testSomethingRelatedToRedis($connection)
{
    $this->app['redis'] = $this->getRedisManager($connection);
    ...
}
```

To get a custom redis connection during a test you can also pass two additional parameters to the `getRedisManager` helper method where the second parameter is the driver and third parameter an connection config array:

```php
$redisManager = $this->getRedisManager('predis_custom', 'predis', [
    'cluster' => false,
    'options' => [
        'prefix' => 'test_',
    ],
    'default' => [
        'url' => "redis://{$host}:{$port}",
        'database' => 5,
        'timeout' => 0.5,
    ],
]);
```

To properly clean up after every test provide the following method call in your tearDown method:

```php
protected function tearDown(): void
{
    $this->tearDownRedis();

    parent::tearDown();
}
```

### CODE REVIEW

File: src/Illuminate/Redis/Connections/PacksPhpRedisValues.php [hunk 1]
```diff
@@ -82,6 +82,17 @@ public function pack(array $values): array
         return array_map($processor, $values);
     }
 
+    /**
+     * Determine if JSON serialization is enabled.
+     *
+     * @return bool
+     */
+    public function jsonSerialized(): bool
```
Comment (taylorotwell) [inline src/Illuminate/Redis/Connections/PacksPhpRedisValues.php:90]: Should this be something like `usesJsonSerialization`?

Comment (TheLevti) [reply src/Illuminate/Redis/Connections/PacksPhpRedisValues.php:90]: Yes, name fits better. Will rename it

Comment (TheLevti) [reply src/Illuminate/Redis/Connections/PacksPhpRedisValues.php]: Renamed it to `usesJsonSerialization`.

**Comment by TheLevti** [comment]
@tillkruss could you have a look as well, please?

**Review by tillkruss — CHANGES_REQUESTED** [review]
This looks good to me. Thank you split-diff-view.

File: src/Illuminate/Foundation/Testing/Concerns/InteractsWithRedis.php [hunk 2]
```diff
@@ -21,97 +24,344 @@ trait InteractsWithRedis
...
+                $config['default']['options']['name'] = 'phpredis_compression_lz4_max';
+                $config['default']['options']['compression'] = Redis::COMPRESSION_LZ4;
+                $config['default']['options']['compression_level'] = 12;
+                break;
+            case 'phpredis_msgpack_and_lz4':
+                $config['default']['options']['name'] = 'phpredis_msgpack_and_lz4';
+                $config['default']['options']['serializer'] = Redis::SERIALIZER_MSGPACK;
+                $config['default']['options']['compression'] = Redis::COMPRESSION_LZ4;
+                $config['default']['options']['compression_level'] = 12;
+                break;
```
[the remark this answers is not in the document]
Comment (TheLevti) [reply src/Illuminate/Foundation/Testing/Concerns/InteractsWithRedis.php:199]: Those are all the edge cases that should be tested that may result in different behaviours. I can remove all the connection specific variations as functionally nothing changes and the connector test has them anyway specifically covered. But everything from scan, we should try as this makes sure that all variations of what phpredis supports (serialization and compression) is covered. There were some cases where e.g. json serializer was causing something to fail. Same for compression.

I will then also remove the min/max level connections as they are not really changing anything of the outcome.

[the remark this answers is not in the document]
Comment (TheLevti) [reply src/Illuminate/Foundation/Testing/Concerns/InteractsWithRedis.php]: Removed them in https://github.com/laravel/framework/pull/41718/commits/71c7ad9e699f1cc2db27225d886eeb4b308a4559

should be ready to be reviewed again.

**Review by tillkruss — APPROVED** [review]

**Comment by taylorotwell** [comment]
I get some Redis failures when running `bash bin/test.sh`

**Comment by TheLevti** [comment]
> I get some Redis failures when running `bash bin/test.sh`

Thank you, I will try to run it locally and fix the remaining issue. So far I was just targeting the CI to pass.

**Comment by TheLevti** [comment]
> I get some Redis failures when running `bash bin/test.sh`

Could you maybe tell me which test fails for you locally? I currently only see that the various redis connection tests fail (scheme, sentinel, username/password). But not sure whether this is because I am running the testsuite on a linux distro rather than mac (test.sh was written for mac it seems).

I will try to resolve those connection issues.

**Comment by TheLevti** [comment]
@taylorotwell @tillkruss The pull request is ready for the final review. Local test run is fixed (I was not aware of /bin/test.sh before 😅).

The issue was that the local setup is not configured to support actual tls/sentinel connections (new code tries to always connect, flush the db on connect and flush the db on disconnect in the end, which does not work locally for those connections). So I reverted those specific connection tests to just validate whether the connection params are properly set.

**Comment by Huggyduggy** [comment]
I've made some adjustments within phpunits config

         cacheResult ="false"
         enforceTimeLimit="true"  (necessary as I don't have dynamodb AWS keys)

    <logging>
    <log type="testdox-text" target="php://stdout"/>
</logging>

And got the following output on W11 + WSL2 (tests running within docker). None of those seem to have any connection to redis tests, but are related to either the test.sh being written with macOS in mind or the docker image itself (webp missing -> gd): [pull41718.log](https://github.com/laravel/framework/files/8822479/pull41718.log)

**Comment by driesvints** [comment]
@TheLevti I see the same thing as Taylor. Works on 9.x but get weird results on your branch.

**Comment by TheLevti** [comment]
> @TheLevti I see the same thing as Taylor. Works on 9.x but get weird results on your branch.

Thanks guys, I will see what I can find. I don't see a reason how it is related to the application code/redis changes.

I have a feeling that we have a bad local test env setup and also as we noticed its tightly coupled to mac users only, which should not be when using docker. Most likely my changes triggered some tests to actually run/write to disk, which did not work properly before. For example for me the suite on my branch only starts to work once I change the owner of all files to www-data (not in the container), which should not be required as we can not expect that everyone has a www-data user.

Going to try and fix this platform inconsistency these days.

**Comment by tillkruss** [comment]
This works for me:

```
gh repo clone laravel/framework
gh pr checkout 41718

vendor/bin/phpunit
```



<details>
<summary>But I do get a couple of incomplete tests:</summary>
<pre><code>
There were 12 incomplete tests:

1) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "predis" ('predis')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

2) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis" ('phpredis')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

3) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_prefix" ('phpredis_prefix')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

4) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_scan_noretry" ('phpredis_scan_noretry')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

5) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_scan_retry" ('phpredis_scan_retry')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

6) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_scan_prefix" ('phpredis_scan_prefix')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

7) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_scan_noprefix" ('phpredis_scan_noprefix')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

8) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_serializer_php" ('phpredis_serializer_php')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

9) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_serializer_igbinary" ('phpredis_serializer_igbinary')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

10) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_serializer_json" ('phpredis_serializer_json')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

11) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_compression_lzf" ('phpredis_compression_lzf')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88

12) Illuminate\Tests\Integration\Cache\RedisStoreTest::testItCanAddWithTtl with data set "phpredis_compression_zstd" ('phpredis_compression_zstd')
Needs support in the RedisStore first.

/Users/Till/Development/Repositories/framework/tests/Integration/Cache/RedisStoreTest.php:88
</code></pre>
</details>

**Comment by TheLevti** [comment]
> 

This is fine. I added those as a preparation for my next pull request. Currently the Cache component does not support any of the serializer/compressors. I want this to be merged first before I submit the next pull request that adds this functionality

**Comment by tillkruss** [comment]
>Currently the Cache component does not support any of the serializer/compressors. I want this test suite improvement to be merged first before I submit the next pull request that adds this functionality

Ah nice, I like where you're going with this!

**Comment by driesvints** [comment]
@TheLevti if we can't fix the tests then for now it's probably best to don't go forward with this.

**Comment by TheLevti** [comment]
@driesvints @tillkruss could you please try again? I updated the test.sh to support any platform. It seems like it was written specifically for mac os only. It should not run on mac os, linux or windows the same way.

In addition I noticed that sometimes windows tests failed, because the `active_url` validator test was doing real DNS lookups, which must be avoided during test runs as sometimes this times out or fails. This should not be fixed as well as the dns call is now mocked in the test.

The added changes are:

- Make sure host gateway is mapped properly so the generic domain name is available on mac os/linux/windows.
- Make sure to mount the file system with the same user as the host so that we do not have inconsistencies when files are written by tests.
- Mock actual DNS lookup calls (dns_get_record).

**Comment by taylorotwell** [comment]
Output is still all messed up for me. I don't really have time atm to take on such a big PR. Sorry 🫤

**Comment by TheLevti** [comment]
> Output is still all messed up for me. I don't really have time atm to take on such a big PR. Sorry 🫤

Can someone please tell me what is wrong with the output? The changes are not related and ci/tests are passing for my proposed changes perfectly fine. I suspect we still have a non os independent local test setup, which causes some trouble.

**Comment by tillkruss** [comment]
Tests run fine for me on macOS 12.4 using iTerm2 3.4.16.

**Comment by driesvints** [comment]
@TheLevti I think the best course of action for now might be to send in smaller incremental PR's to see if we can solve this and spot the output bug in one of them. Like Taylor indicated, it will probably be impossible to review such a large PR.

**Comment by TheLevti** [comment]
> @TheLevti I think the best course of action for now might be to send in smaller incremental PR's to see if we can solve this and spot the output bug in one of them. Like Taylor indicated, it will probably be impossible to review such a large PR.

I would like to avoid losing this changes as they have been apready reviewed by others and the final benefit is worth the effort. But if Tylor has the last say, I will of course try with smaller chuncks if this is the only way.

It would be helpful though to get the locally broken output reproduced though. If anyone could contribute with their output plus terminal they use where it is broken, this would be nice.

My suggestion here then would be:

1) pr to improve local test runs.
2) pr to improve the redis test suite (it cant be a lot smaller as i need to update all test classes that use redis to make sure that all future redis changes are properly tested (using data providers)

Would that work or can we get this pull request fixed somehow together?

**Comment by driesvints** [comment]
@taylorotwell ^

**Comment by tillkruss** [comment]
Yeah, it's hard to make this PR smaller. Maybe someone who gets the wonky output can do a quick call for diagnostics?

**Comment by TheLevti** [comment]
Does it happen to you when you do a fresh git clone and run the tests on your machine @driesvints as for you the output was also broken before? I suspect some old not tracked files/folders with an incorrect user/acl cause some issues.

**Comment by driesvints** [comment]
I can confirm I no longer see the weird test output. I have no idea why that is... I ran `./bin/test/sh` and then `phpunit` and it properly ran the tests for me.

@TheLevti we recently made a move to remove as many skipped output as possible. However, this PR re-adds a bunch of that, making it harder to dig into failed tests in the CI. Can you do this differently maybe?

Screenshot 2022-07-26 at 10 37 20

**Comment by TheLevti** [comment]
> I can confirm I no longer see the weird test output. I have no idea why that is... I ran `./bin/test/sh` and then `phpunit` and it properly ran the tests for me.
> 
> @TheLevti we recently made a move to remove as many skipped output as possible. However, this PR re-adds a bunch of that, making it harder to dig into failed tests in the CI. Can you do this differently maybe?
> 
> Screenshot 2022-07-26 at 10 37 20

I added those tests, because I wanted to add the necessary support in the follow up pull request where those tests would run successfully. Shall I leave them and wait for the next pull request or get them out? My follow up pull request will add proper support on the cache component. I just first wanted to clean up redis tests.

**Comment by driesvints** [comment]
It's best to filter out this before we merge. Otherwise it's going to be annoying to review other pr's or failing builds where test fail.

**Comment by TheLevti** [comment]
> It's best to filter out this before we merge. Otherwise it's going to be annoying to review other pr's or failing builds where test fail.

Ok, sure. Will re-add them once the relevant code will be added. Thank you for the input 🥳

**Comment by TheLevti** [comment]
@driesvints Incomplete tests are gone. What are the next steps?

**Comment by driesvints** [comment]
@taylorotwell will review this next

**Comment by taylorotwell** [comment]
I still get a bunch of errors locally - I have no idea. I again don't really have a ton of time to dig into this and it's a huge change to our test setup for Redis that I won't know anything about. Is there a particular bug you want to fix in Redis that exists in Laravel right now? If so, can we just fix that and add a test for it?

**Comment by TheLevti** [comment]
> I still get a bunch of errors locally - I have no idea. I again don't really have a ton of time to dig into this and it's a huge change to our test setup for Redis that I won't know anything about. Is there a particular bug you want to fix in Redis that exists in Laravel right now? If so, can we just fix that and add a test for it?

No, this is not a bug fix, but a cleanup for the current redis tests, because they are not well structured and each component that uses redis has a non consistent test flow.

This pull request would first make sure that everything that uses redis, would use the same data providers. Its all about that.

I wanted to do this to make sure that the planned follow up pull requests will make sure that everything still works.

I will then just provide the follow up pull request with just the necessary tests.
